processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
+        // further compute for child nodes
+        sum += computeRequiredBufferNum(Math.min(numNodesPerChild, totalNumNodes - i * numNodesPerChild), degree);
+      }
+
+      return sum;
+    } else {
+      return 1;
+    }
+  }
+
+  /**
+   * Recursively build a combining tree in a top-down manner.
+   *
+   * @param sortedIterators    sorted iterators
+   * @param bufferSupplier     combining buffer supplier
+   * @param combiningFactories array of combining aggregator factories
+   * @param combineDegree      combining degree
+   *
+   * @return an iterator of the root of the combining tree
+   */
+  private Iterator<Entry<KeyType>> buildCombineTree(
+      List<Iterator<Entry<KeyType>>> sortedIterators,
+      Supplier<ByteBuffer> bufferSupplier,
+      AggregatorFactory[] combiningFactories,
+      int combineDegree,
+      List<Future> combineFutures,
+      List<String> dictionary
+  )
+  {
+    final int numIterators = sortedIterators.size();
+    if (numIterators > combineDegree) {
+      final int iteratorsPerChild = (numIterators + combineDegree - 1) / combineDegree; // ceiling
+      final List<Iterator<Entry<KeyType>>> childIterators = new ArrayList<>();
+      for (int i = 0; i < combineDegree; i++) {
+        childIterators.add(
+            buildCombineTree(
+                sortedIterators.subList(i * iteratorsPerChild, Math.min(numIterators, (i + 1) * iteratorsPerChild)),
+                bufferSupplier,
+                combiningFactories,
+                combineDegree,
+                combineFutures,
+                dictionary
+            )
+        );
+      }
+      return runCombiner(childIterators, bufferSupplier.get(), combiningFactories, combineFutures, dictionary).iterator();
+    } else {
+      return runCombiner(sortedIterators, bufferSupplier.get(), combiningFactories, combineFutures, dictionary).iterator();
+    }
+  }
+
+  private static ByteBuffer getSlice(ByteBuffer buffer, int sliceSize, int i)
+  {
+    final ByteBuffer slice = buffer.duplicate();
+    slice.position(sliceSize * i);
+    slice.limit(slice.position() + sliceSize);
+    return slice.slice();
+  }
+
+  private StreamingMergeSortedGrouper<KeyType> runCombiner(
+      List<Iterator<Entry<KeyType>>> iterators,
+      ByteBuffer combineBuffer,
+      AggregatorFactory[] combiningFactories,
+      List<Future> combineFutures,
+      List<String> dictionary
+  )
+  {
+    final Iterator<Entry<KeyType>> mergedIterator = Groupers.mergeIterators(iterators, keyObjComparator);
+    final SettableColumnSelectorFactory settableColumnSelectorFactory =
+        new SettableColumnSelectorFactory(aggregatorFactories);
+    final StreamingMergeSortedGrouper<KeyType> grouper = new StreamingMergeSortedGrouper<>(
+        Suppliers.ofInstance(combineBuffer),
+        combineKeySerdeFactory.factorizeWithDictionary(dictionary),
+        settableColumnSelectorFactory,
+        combiningFactories
+    );
+    grouper.init();
```

Comment:
Why not calling this inside lambda

Reply:
Moved.

Reply:
This should be called before outside because it initialize some variables like curWriteIndex, nextReadIndex, and finished which should be done before calling iterator(). I reverted this change.